tests/modcc/test_parser.cpp

```diff
@@ -15,481 +79,413 @@ TEST(Parser, full_file) {
+    }
 
-#ifdef VERBOSE_TEST
-        if(e) std::cout << e->to_string() << std::endl;
-#endif
-        EXPECT_NE(e, nullptr);
-        EXPECT_EQ(p.status(), lexerStatus::happy);
+    const char* double_expr[] = {
+        "B+A", "a1 + 2bn", "4c+d"
+    };
 
-        // always print the compiler errors, because they are unexpected
-        if(p.status()==lexerStatus::error)
-            std::cout << cyan(expression) << "\t"
-                      << red("error") << p.error_message() << std::endl;
+    for (auto& text: double_expr) {
+        std::unique_ptr<StoichExpression> s;
+        EXPECT_TRUE(check_parse(s, &Parser::parse_stoich_expression, text));
+        EXPECT_EQ(2, s->terms().size());
     }
 
-    std::vector<const char*> bad_expressions =
-    {
-"(x             ",
-"((x+3)         ",
-"(x+ +)         ",
-"(x=3)          ",  // assignment inside parenthesis isn't allowed
-"(a + (b*2^(x)) ",  // missing closing parenthesis
+    const char* other_good_expr[] = {
+        "", "a+b+c", "1a-2b+3c+4d"
     };
 
-    for(auto const& expression : bad_expressions) {
-        Parser p(expression);
-        auto e = p.parse_parenthesis_expression();
+    for (auto& text: other_good_expr) {
+        std::unique_ptr<StoichExpression> s;
+        EXPECT_TRUE(check_parse(s, &Parser::parse_stoich_expression, text));
+    }
 
-        EXPECT_EQ(e, nullptr);
-        EXPECT_EQ(p.status(), lexerStatus::error);
+    const char* check_coeff = "-3a+2b-c+d";
+    {
+        std::unique_ptr<StoichExpression> s;
+        EXPECT_TRUE(check_parse(s, &Parser::parse_stoich_expression, check_coeff));
+        EXPECT_EQ(4, s->terms().size());
+        std::vector<int> confirm = {-3,2,-1,1};
+        for (unsigned i = 0; i<4; ++i) {
+            auto term = s->terms()[i]->is_stoich_term();
+            EXPECT_EQ(confirm[i], term->coeff()->is_integer()->integer_value());
+        }
+    }
 
-#ifdef VERBOSE_TEST
-        if(e) std::cout << e->to_string() << std::endl;
-        if(p.status()==lexerStatus::error)
-            std::cout << "in " << cyan(expression) << "\t" << p.error_message() << std::endl;
-#endif
+    const char* bad_expr[] = {
+        "A+B+", "A+5+B"
+    };
+
+    for (auto& text: bad_expr) {
+        EXPECT_TRUE(check_parse_fail(&Parser::parse_stoich_expression, text));
     }
 }
 
-// test parsing of line expressions
-TEST(Parser, parse_line_expression) {
-    std::vector<const char*> good_expressions =
-    {
-"qt=q10^((celsius-22)/10)"
-"x=2        ",
-"x=2        ",
-"x = -y\n   "
-"x=2*y      ",
-"x=y + 2 * z",
-"x=(y + 2) * z      ",
-"x=(y + 2) * z ^ 3  ",
-"x=(y + 2 * z ^ 3)  ",
-"foo(x+3, y, bar(21.4))",
-"y=exp(x+3) + log(exp(x/y))",
-"a=x^y^z",
-"a=x/y/z"
+// test parsing of stoich and reaction expressions
+TEST(Parser, parse_reaction_expression) {
+    const char* good_expr[] = {
+        "~ A + B <-> C + D (k1, k2)",
+        "~ 2B <-> C + D + E (k1(3,v), k2)",
+        "~ <-> C + D + 7 E (k1, f(a,b)-2)",
```

Comment:
Here you have a space between the number and the `E`. We discussed this didn't we, and decided that the compiler error message about an unexpected real number would be enough, didn't we?